classes/booking_manager.php

```diff
@@ -0,0 +1,316 @@
+        $this->facetoface = $facetoface;
+        $this->course = $course;
+        $this->records = $records;
+    }
+
+    /**
+     * Returns file from file system. File must exist.
+     * @param int $fileitemid Item id of file stored in the current $USER's draft file area
+     * @return stored_file
+     */
+    public function load_from_file(int $fileitemid) {
+        global $USER;
+        $this->usefile = true;
+
+        $fs = new file_storage();
+        $files = $fs->get_area_files(context_user::instance($USER->id)->id, 'user', 'draft', $fileitemid, 'itemid', false);
+
+        if (count($files) != 1) {
+            throw new moodle_exception('error:cannotloadfile', 'mod_facetoface');
+        }
+
+        return current($files);
+    }
+
+    /**
+     * Load in the records to process from an array
+     * @param array $records
+     */
+    public function load_from_array(array $records) {
+        $this->usefile = false;
+        $this->records = $records;
+
+        return $this;
+    }
+
+    public static function get_headers(): array {
+        return [
+            'email',
+            'session',
+            'status',
+            'discountcode',
+            'notificationtype',
+        ];
+    }
+
+    private function get_iterator() {
+        if (!$this->usefile) {
+            foreach ($this->records as $record) {
+                yield $record;
+            }
+            return;
+        }
+
+        $handle = $this->file->get_content_file_handle();
+        $maxlinelength = 1000;
+        $delimiter = ',';
+        $rownumber = 1; // First row is headers.
+        $headers = self::get_headers();
+        $numheaders = count($headers);
+        fgets($handle); // Move pointer past first line (headers).
+        try {
+            while (($data = fgetcsv($handle, $maxlinelength, $delimiter)) !== false) {
+                $rownumber++;
+                $numfields = count($data);
+                if ($numfields !== $numheaders) {
+                    throw new moodle_exception('error:bookingsuploadfileheaderfieldmismatch', 'mod_facetoface');
+                }
+                $record = array_combine($headers, $data);
+                yield (object) $record;
+            }
+        } finally {
+            fclose($handle);
+        }
+    }
+
+    public function validate() {
+        global $DB;
+        $errors = [];
+        $sessioncapacitycache = [];
+
+        // Break into rows.
+        foreach ($this->get_iterator() as $index => $entry) {
+            $row = $index + 1;
+
+            // Set defaults for fields with no value.
+            $entry->status = $entry->status ?? '';
+            $entry->notificationtype = $entry->notificationtype ?? '';
+            $entry->discountcode = $entry->discountcode ?? '';
+
+            // Validate and get user.
+            $userids = $DB->get_records('user', ['email' => $entry->email], 'id');
+
+            // Multiple matched, ambiguous which is the real one.
```

Comment:
Given that these have a couple of conditions on most of the fields, I reckon we should just split this up into a couple of private functions that just take a field -> errors, or row -> errors if there is cross dependencies. that way the main loop just reads as the series of checks being done at a high level

Reply:
I tested for user, and that was okay.., being that it worked when I added a method like `check_user($email, &$errors, $row) -> $userid`, but the problem is it introduced some level of inconsistency in how the checks are performed and recorded compared to sessions, which required different parameters to be provided, along with a session cache object, so it's probably better to keep all the checks inline for now, until there's a better path forward for all the checks collectively.

Reply:
For now I've left it inline and added a comment to the function (which was missing at the time)